File: BinarySearch/Upper_Bound.cpp
#include<iostream>
using namespace std;

/* Upper Bound Function
 * Returns the least (first) index of the smallest element in the array strictly greater than key
 * If the key is greater than the maximum element of the array, it returns -1
 */
int upper_bound(int arr[], int lo, int hi, int key) {
	int mid, ans = hi+1;
	while(lo <= hi) {
		mid = lo + (hi-lo)/2;
		if(arr[mid] > key) {
			ans = mid;
			hi = mid-1;
		}
		else
			lo = mid+1;
	}
	return ans;
}

int main() {
	int n;
	cin >> n;
	int a[n];
	for(int i=0;i<n;i++)
		cout << i << " ";
	cout << endl;
	for(int i=0;i<n;i++)
		cin >> a[i];
	int k;
	cin >> k;
	int ub = upper_bound(a, 0, n-1, k);
	if(ub != n)
		cout << a[ub] << " at position: " << ub << "\n";
	else
		cout << "Does not exist in this array\n";
	return 0;
}

File: Java/src/BinarySearch/UpperBound.java
package BinarySearch;

import FastScanner.FastScanner;

import java.io.BufferedOutputStream;
import java.io.PrintWriter;

/*
    Find the first element in array that is strictly greater than the key
    Return -1 if such an element does not exist
 */
public class UpperBound {
    static PrintWriter out;
    
    private static int upperBound(int[] a, int key) {
        int lo = 0, hi = a.length-1, mid, ans = -1;
        while(lo <= hi) {
            mid = lo + (hi-lo)/2;
            if(a[mid] > key) {
                ans = mid;
                hi = mid-1;
            } else {
                lo = mid+1;
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        FastScanner fs = new FastScanner();
        out = new PrintWriter(new BufferedOutputStream(System.out));
        
        int n = fs.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = fs.nextInt();
        }
        int x = fs.nextInt();
        
        out.println(upperBound(a, x));
        
        out.close();
    }
}
